tsdb/head_wal.go

```diff
@@ -193,122 +194,127 @@ func (h *Head) loadWAL(r *wal.Reader, multiRef map[chunks.HeadSeriesRef]chunks.H
 	// The records are always replayed from the oldest to the newest.
 Outer:
 	for d := range decoded {
-		switch v := d.(type) {
-		case []record.RefSeries:
-			for _, walSeries := range v {
-				mSeries, created, err := h.getOrCreateWithID(walSeries.Ref, walSeries.Labels.Hash(), walSeries.Labels)
-				if err != nil {
-					seriesCreationErr = err
-					break Outer
-				}
+		select {
+		case <-ctx.Done():
+			break
```

Comment:
We should be returning an error in this case. If it was the last file being replayed, then it will be silently ignored.
```
seriesCreationErr = ErrWALReplayAborted
break Outer
```

Reply:
Added errors, but I cannot re-use `seriesCreationErr` there since it's being written to from the loop below and so race detector (correctly) complains about multiple goroutines writing to the same error. Added a dedicated error variable for this.